src/main/java/bio/terra/common/tracing/TracingHook.java

```diff
@@ -0,0 +1,214 @@
+    // End the Flight Span for the current run.
+    flightSpan.end();
+    return HookAction.CONTINUE;
+  }
+
+  @Override
+  public Optional<StepHook> stepFactory(FlightContext context) {
+    return Optional.of(new TraceStepHook(flightSpans.get(context.getFlightId())));
+  }
+
+  /** A {@link StepHook} for creating Spans for each Step execution. */
+  private class TraceStepHook implements StepHook {
+    private final Span flightSpan;
+    private Span stepSpan = null;
+    private Scope stepScope = null;
+
+    public TraceStepHook(Span flightSpan) {
+      this.flightSpan = flightSpan;
+    }
+
+    @Override
+    public HookAction startStep(FlightContext flightContext) {
+      // Start the Step Span. We must remember to close the Span at the end of the Step.
+      stepSpan =
+          Tracing.getTracer()
+              .spanBuilderWithExplicitParent(
+                  STEP_NAME_PREFIX + flightContext.getStepClassName(), flightSpan)
+              .startSpan();
+      stepSpan.putAttribute(
+          "stairway/flightId", AttributeValue.stringAttributeValue(flightContext.getFlightId()));
+      stepSpan.putAttribute(
+          "stairway/flightClass",
+          AttributeValue.stringAttributeValue(flightContext.getFlightClassName()));
+      stepSpan.putAttribute(
+          "stairway/stepClass",
+          AttributeValue.stringAttributeValue(flightContext.getStepClassName()));
+      stepSpan.putAttribute(
+          "stairway/direction",
+          AttributeValue.stringAttributeValue(flightContext.getDirection().toString()));
+      // Start the Scope of the Step Span's execution so that the step execution has a relevant
+      // current span. We must remember to close the Scope at the end of the Step.
+      stepScope = tracer.withSpan(stepSpan);
+      return HookAction.CONTINUE;
+    }
+
+    @Override
+    public HookAction endStep(FlightContext flightContext) {
+      stepScope.close();
+      stepSpan.end();
+      return HookAction.CONTINUE;
+    }
+  }
+
+  /**
+   * Gets or creates the context of the submission Span. Each Flight span is linked to the
+   * submission span to correlate all Flight Spans.
+   */
+  private SpanContext getOrSubmissionContext(FlightContext flightContext) {
+    // Check the input parameters for a submission span to link for the Flight.
+    String encodedContext =
+        flightContext.getInputParameters().get(SUBMISSION_SPAN_CONTEXT_MAP_KEY, String.class);
+    if (encodedContext != null) {
+      return decodeContext(encodedContext);
+    }
+    // Check the working map for a submission span to link for the Flight.
+    encodedContext =
+        flightContext.getWorkingMap().get(SUBMISSION_SPAN_CONTEXT_MAP_KEY, String.class);
+    if (encodedContext != null) {
+      return decodeContext(encodedContext);
+    }
+    // Create a new span to use as a submission span to link for the Flight Spans.
+    Span submissionSpan =
+        tracer
+            .spanBuilderWithExplicitParent(
```

Comment:
We already serelize/deserelize spanContext in flight Map.
if we use [spanBuilderWithRemoteParent](https://www.javadoc.io/doc/io.opencensus/opencensus-api/0.17.0/io/opencensus/trace/Tracer.html#spanBuilderWithRemoteParent-java.lang.String-io.opencensus.trace.SpanContext-)
Can we also get correct span even in fligjt recover case?

Reply:
Sorry, I'm not sure what you mean.
We do not want the Flight Spans to be the children of the Submission span because the Flight Spans may outlive the Submission Span. Child spans must be completed before their parent spans. With the parent *link*, we can have the "child linked" spans outlive their "parent linked" spans. (Terminology around linkage is confusing, but that's what they have. https://opencensus.io/tracing/span/link/)

I don't know what you mean by the flight recover case. The same Flight Span is in use when the Flight is undone on error.

Reply:
Sorry for the confusion, the recovery I mean is this scenario: Flight running, app is redeployed. Then we lost the original `flightSpan`. Do you think it make sense if we reconstruct the `remoteParentSpanContext` ourself.
Just my through, totally optional.

Reply:
We can't resume a Span, even with a SpanContext. So if the app crashes ungracefully, and we have dangling Spans, they will not get exported. See my longer answer to Greg about why we can't have a single Span for the entire Flight's lifetime across Stairway instances.